File: solutions/2816. Double a Number Represented as a Linked List/2816_2.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
    if head.val >= 5:
      head = ListNode(0, head)

    curr = head

    while curr:
      curr.val *= 2
      curr.val %= 10
      if curr.next and curr.next.val >= 5:
        curr.val += 1
      curr = curr.next

    return head      

File: solutions/2816. Double a Number Represented as a Linked List/2816_2.rs
// Time complexity: O(n)
// Space complexity: O(1)
impl Solution {
    pub fn double_it(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head = Some(Box::new(ListNode { val: 0, next: head }));
        let mut prev = head.as_mut()?;
        while let Some(digit) = prev.next.as_mut() {
            prev.val += (digit.val * 2) / 10;
            digit.val = (digit.val * 2) % 10;
            prev = digit;
        }
        if head.as_ref().is_some_and(|v| v.val == 0) {
            head = head?.next.take();
        }

        head
    }
}
